cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/client/AzureClientCredentials.java

```diff
@@ -92,45 +93,70 @@ public MarketplaceOrderingManager getMarketplaceOrderingManager() {
-                String appReplyUrl = credentialView.getAppReplyUrl();
-                String authorizationCode = credentialView.getAuthorizationCode();
-                result = new CbDelegatedTokenCredentials(applicationTokenCredentials, appReplyUrl, authorizationCode, secretKey, authenticationContextProvider,
-                        cbRefreshTokenClientProvider);
-            }
+        if (credentialView.codeGrantFlow()) {
+            result = delegatedTokenCredentials(azureEnvironment, applicationTokenCredentials);
         } else {
             LOGGER.info("Creating Azure credentials with application token credentials, credential: {}", credentialView.getName());
         }
         return result.withDefaultSubscriptionId(subscriptionId);
     }
 
+    private ApplicationTokenCredentials certificateBasedCredentials(AzureEnvironment azureEnvironment) {
+        LOGGER.info("Creating Azure credentials with certificate and private key: {}", credentialView.getName());
+        String auth = credentialView.getPrivateKeyForCertificate() + credentialView.getCertificate();
+        return new ApplicationTokenCredentials(credentialView.getAccessKey(), credentialView.getTenantId(),
+                auth.getBytes(), null, azureEnvironment);
+    }
+
+    private ApplicationTokenCredentials secretBasedCredentials(AzureEnvironment azureEnvironment) {
+        LOGGER.info("Creating Azure credentials with secret: {}", credentialView.getName());
+        return new ApplicationTokenCredentials(credentialView.getAccessKey(), credentialView.getTenantId(),
+                credentialView.getSecretKey(), azureEnvironment);
+    }
+
+    private AzureTokenCredentials delegatedTokenCredentials(AzureEnvironment azureEnvironment, ApplicationTokenCredentials applicationTokenCredentials) {
+        // Probably this is a dead code at the moment, since code grant flow is not supported, I just put it into a separate method
+        String tenantId = credentialView.getTenantId();
+        String clientId = credentialView.getAccessKey();
+        String secretKey = credentialView.getSecretKey();
+        AzureTokenCredentials result;
+        String refreshToken = credentialView.getRefreshToken();
+        if (StringUtils.isNotEmpty(refreshToken)) {
+            LOGGER.info("Creating Azure credentials for a new delegated token with refresh token, credential: {}", credentialView.getName());
+            String resource = azureEnvironment.managementEndpoint();
+            CBRefreshTokenClient refreshTokenClient = cbRefreshTokenClientProvider.getCBRefreshTokenClient(azureEnvironment.activeDirectoryEndpoint());
+            AuthenticationResult authenticationResult = refreshTokenClient.refreshToken(tenantId, clientId, secretKey, resource, refreshToken, false);
+
+            if (authenticationResult == null) {
+                String msg = String.format("New token couldn't be obtain with refresh token for credential: %s", credentialView.getName());
```

Comment:
CAn we somehow propagate a more userfriendly (easy to understand ) message to the user? e.g include what should be done by them to avoid this error.

Reply:
It is not a new code it was just a move. I am happy to delete it, refactoring doe snot make sense since it is not possible to test, since it is a dead code.